<a href="https://colab.research.google.com/github/sw01-prog/python_coding_practice/blob/main/challenge_set_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Fibonacci series

It is used to find the 7th indexed number in the fibonacci series

In [ ]:
n=int(input())

7


In [ ]:
def fibonacci(n):
 if n==1 or n==2:
  return 1
 else:
  return fibonacci(n-1)+fibonacci(n-2)

print(fibonacci(n))

13


Sum of even and odd numbers

In [ ]:
l1=[1,2,3,4,5,6,7,8]
even_sum =0
odd_sum = 0

def sum_even_odd(l1):
 even_sum =0
 odd_sum = 0
 for i in l1:
  if i%2==0:
   print(even_sum+=i)


  else:
   print(odd_sum+=i)


sum_even_odd(l1) # call the function to execute it

SyntaxError: invalid syntax (ipython-input-2029275473.py, line 10)

In [ ]:
l1=[1,2,3,4,5,6,7,8]
#even_sum =0
#odd_sum = 0

def sum_even_odd(l1):
 even_sum =0
 odd_sum = 0
 for i in l1:
  if i%2==0:
   even_sum+=i
  elif i%2!=0:
   odd_sum+=i

 return(even_sum,odd_sum)

even_sum, odd_sum = sum_even_odd(l1)
print(even_sum, odd_sum)


UnboundLocalError: cannot access local variable 'odd_sum' where it is not associated with a value

the error happens because of Python’s variable scope rules.

Inside your function sum_even_odd,

When you write even_sum += i or odd_sum += i,
Python assumes even_sum and odd_sum are local variables to the function.

But you never created them inside the function, so they are “unbound” → hence UnboundLocalError.

In [ ]:
l1=[1,2,3,4,5,6,7,8]
#even_sum =0
#odd_sum = 0

def sum_even_odd(l1):
 even_sum =0
 odd_sum = 0
 for i in l1:
  if i%2==0:
   even_sum+=i
  elif i%2!=0:
   odd_sum+=i

 return(even_sum,odd_sum)

even_sum, odd_sum = sum_even_odd(l1)
print(even_sum, odd_sum)

20 16


Palindrome number

In [ ]:
l1=[1,2,1]

def check_palindrome(l1):
 string_m= ''.join(l1)
 if string_m == string_m[::-1]:
  print('Palindrome')
 else:
  print('Not Palindrome')
check_palindrome(l1)

TypeError: sequence item 0: expected str instance, int found

The error happens because ''.join() only works on sequences of strings, but your list l1 contains integers.



In [ ]:
l1 = [1, 2, 1]

def check_palindrome(l1):
    # Convert all integers to strings first
    string_m = ''.join(map(str, l1))
    if string_m == string_m[::-1]:
        print('Palindrome')
    else:
        print('Not Palindrome')

check_palindrome(l1)


Palindrome


one more thing to check

In [ ]:
l1 = [1, 2, 1]

def check_palindrome(l1):
    # Convert all integers to strings first
    string_m = ''.join(map(str, l1))
    if string_m == string_m[::-1]:
        return True
    else:
        return False

print(check_palindrome(l1))

True


In [ ]:
l1 = [1, 2, 1]

def check_palindrome(l1):
    # Convert all integers to strings first
    string_m = ''.join(map(str, l1))
    if string_m == string_m[::-1]:
        return True
    else:
        return False

check_palindrome(l1)

True

string reverse

In [ ]:
n= input()
def reverse(n):
 rev= int(str(n)[::-1])  # int() removes leading zeros
 return rev
reverse(n)

4678


8764

Calculator based on conditions

In [ ]:
n= int(input())

def simple_calculator(n):

 for _ in range(n):
  op= int(input())

  if op==6:
   break

  if op>=1 and op<=5:
   a=int(input())
   b=int(input())

   if op==1:
    print(a+b)
   elif op==2:
    print(a-b)
   elif op==3:
    print(a*b)
   elif op==4:
    print(a//b)
   elif op==5:
    print(a%b)

  else:
   print('invalid')
   break

#simple_calculator(n)

3


In [ ]:

simple_calculator(n)

In [ ]:
def simple_calculator(n):
    for _ in range(n):
        op = int(input())  # Operation code

        if op == 6:
            break  # Exit immediately

        if 1 <= op <= 5:
            a = int(input())
            b = int(input())

            if op == 1:
                print(a + b)
            elif op == 2:
                print(a - b)
            elif op == 3:
                print(a * b)
            elif op == 4:
                print(a // b)  # Integer division
            elif op == 5:
                print(a % b)
        else:
            print("Invalid Operation")


# Example usage
n = int(input())  # Number of operations
simple_calculator(n)


1
2
4
4
0


In [ ]:
# Number of operations
n = int(input())

for _ in range(n):
    op = int(input())  # Operation code

    if op == 6:
        break  # Exit the program

    if op >= 1 and op <= 5:
        a = int(input())
        b = int(input())

        if op == 1:
            print(a + b)
        elif op == 2:
            print(a - b)
        elif op == 3:
            print(a * b)
        elif op == 4:
            print(a // b)  # Integer division
        elif op == 5:
            print(a % b)

    else:
        print("Invalid Operation")


4
1
2
4
6


KeyboardInterrupt: Interrupted by user

In [ ]:
n = int(input())
for _ in range(n):
    choice = int(input())
    if choice == 1:
        num1 = int(input())
        num2 = int(input())
        print(num1 + num2)
    elif choice == 2:
        num1 = int(input())
        num2 = int(input())
        print(num1 - num2)
    elif choice == 3:
        num1 = int(input())
        num2 = int(input())
        print(num1 * num2)
    elif choice == 4:
        num1 = int(input())
        num2 = int(input())
        # To handle potential division by zero, though constraints
        # might imply this won't happen.
        if num2 != 0:
            print(num1 // num2)
        else:
            print("Error: Division by zero")
    elif choice == 5:
        num1 = int(input())
        num2 = int(input())
        # To handle potential modulo by zero.
        if num2 != 0:
            print(num1 % num2)
        else:
            print("Error: Modulo by zero")
    elif choice == 6:
        break
    else:
        print("Invalid Operation")

3
1
2
4
6
2
1
3
-2
2
6
7
-1


Sol provided by manshu based on calculator based on condition

In [ ]:
def solution ():
   option = int(input())
   if option==6:
       return
   if option>=7:
       print("Invalid Operation")
       return solution()


   num1 = int(input())
   num2 = int(input())


   if option==1:
       print(num1+num2)
   elif option==2:
       print(num1-num2)
   elif option==3:
       print(num1*num2)
   elif option==4:
       print(num1//num2)
   elif option==5:
       print(num1%num2)


   return solution()
solution()


Convert fareinheit to celsius

In [ ]:
s= int(input())
e= int(input())
w= int(input())

def convrt_to_c(F):
 return((F-32)*5/9)

for F in range(s, e+1, w):
  print(convrt_to_c(F))

0
100
20
-17.77777777777778
-6.666666666666667
4.444444444444445
15.555555555555555
26.666666666666668
37.77777777777778


In [ ]:
S = int(input())
E = int(input())
W = int(input())

def convrt_to_c(F):
    return int((F - 32) * 5 / 9)

for F in range(S, E + 1, W):
    print(F, convrt_to_c(F))


0
100
20
0 -17
20 -6
40 4
60 15
80 26
100 37


lambda is one liner anonymous function

we defined a lambda function(upper) to convert a string to its upper case using upper().

In [ ]:
s1 = 'geeksforgeeks'
s2= lambda func: func.upper()
print(s2(s1))

GEEKSFORGEEKS


lambda with Condition Checking
- # Example: Check if a number is positive, negative, or zero

In [ ]:
z= lambda x: if x>0 print('positive') else print('negative')

SyntaxError: invalid syntax (ipython-input-1359977569.py, line 1)

In [ ]:
z= lambda x: 'positive' if x>0 else 'negative'

In [ ]:
z = lambda x: str(x)[::-1]

In [ ]:
print(z(234678))

876432


Lambda with List Comprehension

In [ ]:
z = [lambda arg=x: arg*10 for x in range(1, 5)]
for i in z:
  print(i())

10
20
30
40


In [ ]:
# Example: Check if a number is even or odd
check = lambda x: "Even" if x % 2 == 0 else "Odd"

print(check(4))
print(check(7))

Even
Odd


In [ ]:
# Example: Perform addition and multiplication in a single line
calc = lambda x, y: (x + y, x * y)

res = calc(3, 4)
print(res)

(7, 12)


if the strings has too much space between words or lettrs how to cut down

In [ ]:
text = "Hello     this    has     too   many spaces"
f = " ".join(text.split())
print(f)

Hello this has too many spaces


In [ ]:
text = "H e l l o,   W o r l d"
f= ' '.join(text.split())
print(f)

H e l l o, W o r l d


T check if a string is emty


In [ ]:
s= ''
if len(s)==0:
  print('empty')
else:
  print('not')

empty


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
4444
4444
4444
4444

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
7
Sample Output 1:
7777777
7777777
7777777
7777777
7777777
7777777
7777777
Sample Input 1:
6
Sample Output 1:
666666
666666
666666
666666
666666
666666


In [2]:
n= int(input())
for i in range(n):
 print(str(n)*n)

4
4444
4444
4444
4444


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
*
**
***
****

Note : There are no spaces between the stars (*).
Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
5
Sample Output 1:
*
**
***
****
*****
Sample Input 2:
6
Sample Output 2:
*
**
***
****
*****
******


In [6]:
n= int(input())
for i in range(1, n+1):
  print('*'* i)

4
*
**
***
****


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
22
333
4444

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
5
Sample Output 1:
1
22
333
4444
55555
Sample Input 2:
6
Sample Output 2:
1
22
333
4444
55555
666666


In [7]:
n= int(input())
for i in range(1, n+1):
  print(str(i)*i)

4
1
22
333
4444


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
11
121
1221

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
5

Sample Output :
1
11
121
1221
12221


In [6]:
n= int(input())
for i in range(1, n+1):
  if i ==1:
    print(1)
  elif i ==2:
   print(11)
  else:
    print(str('1'+'2'*(i-2)+'1'))

6
1
11
121
1221
12221
122221


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
21
321
4321

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
5
Sample Output 1:
1
21
321
4321
54321
Sample Input 2:
6
Sample Output 2:
1
21
321
4321
54321


In [2]:
n= int(input())
for i in range(1, n+1):
 if i ==1:
  print(1)
 if i ==2:
   print(i+str(i-1))

2
1


TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [5]:
n= int(input())
for i in range(1, n+1):
  for j in range(i, 0, -1):
    print(j, end="")
  print()


7
1
21
321
4321
54321
654321
7654321


Step 2: What does print(j, end="") mean?
Normally, print(j) adds a newline after each number.

But end="" tells Python not to add a newline, just print numbers side by side.

But since end="" is used, it does not move to the next line after finishing the row.

So after finishing, the cursor is still at the end of the same line.



This empty print() just moves the cursor to the next line.

Without it, all rows would get printed on the same line.



Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
A
BC
CDE
DEFG

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 13

Sample Input 1:
5

Sample Output 1:
A
BC
CDE
DEFG
EFGHI

Sample Input 2:
6

Sample Output 2:
A
BC
CDE
DEFG
EFGHI
FGHIJK



In [7]:
n = int(input())
for i in range(1, n+1):
 row= ""
 start= 65+ (i-1)
 for j in range(i):
  row +=chr(start+j)
 print(row)

8
A
BC
CDE
DEFG
EFGHI
FGHIJK
GHIJKLM
HIJKLMNO


🔹 Step-by-step Explanation
1. Input
python
Copy
Edit
N = int(input())
Reads the number of rows.

Example: If N = 4, we want 4 rows.

2. Outer loop (rows)
python
Copy
Edit
for i in range(1, N+1):
Runs from 1 to N.

Controls which row we are printing.

So for N=4: i = 1, 2, 3, 4.

3. Initialize row string
python
Copy
Edit
row = ""
We’ll build each row as a string before printing.

4. Find the starting character of the row
python
Copy
Edit
start = 65 + (i - 1)
Recall: chr(65) = 'A'.

So 65 + (i-1) gives the ASCII code of the row’s first letter.

Example:

Row 1 → 65 + (1-1) = 65 → 'A'

Row 2 → 65 + (2-1) = 66 → 'B'

Row 3 → 65 + (3-1) = 67 → 'C'

Row 4 → 65 + (4-1) = 68 → 'D'

5. Inner loop (characters in a row)
python
Copy
Edit
for j in range(i):
    row += chr(start + j)
This loop adds characters to the row.

It runs exactly i times (so row i has i characters).

Each time, it takes the next alphabet using chr(start + j).

Example (Row 3, i=3):

start = 67 ('C')

Iterations:

j=0 → chr(67+0) = 'C'

j=1 → chr(67+1) = 'D'

j=2 → chr(67+2) = 'E'

Row string becomes "CDE".

6. Print the row
python
Copy
Edit
print(row)
After building one row, we print it.

Then outer loop moves to the next row.



Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 5
E
DE
CDE
BCDE
ABCDE

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 26

Sample Input 1:
8

Sample Output 1:
H
GH
FGH
EFGH
DEFGH
CDEFGH
BCDEFGH
ABCDEFGH
Sample Input 2:
7
Sample Output 2:
G
FG
EFG
DEFG
CDEFG
BCDEFG
ABCDEFG



In [10]:
n= int(input())
for i in range(1, n+1):
 start = 65 + (n-i)
 end = 65+n
 row= "".join(chr(ch) for ch in range(start, end))
 print(row)

6
F
EF
DEF
CDEF
BCDEF
ABCDEF


🔹 Step-by-Step Breakdown
1. Input
python
Copy
Edit
N = int(input())
Reads the number of rows.

Example: If N = 5, we want 5 rows.

2. Outer loop
python
Copy
Edit
for i in range(1, N+1):
Runs from 1 to N.

Controls which row we are printing.

So if N=5, i = 1,2,3,4,5.

3. Find starting letter
python
Copy
Edit
start = 65 + (N - i)
chr(65) → 'A', chr(66) → 'B', …, chr(90) → 'Z'.

So 65 is ASCII code of 'A'.

For each row, we calculate the ASCII code of the first character.

Example when N=5:

Row 1 (i=1): 65 + (5-1) = 69 → 'E'

Row 2 (i=2): 65 + (5-2) = 68 → 'D'

Row 3 (i=3): 65 + (5-3) = 67 → 'C'

Row 4 (i=4): 65 + (5-4) = 66 → 'B'

Row 5 (i=5): 65 + (5-5) = 65 → 'A'

✅ That matches the expected starting letters.

4. Find ending letter
python
Copy
Edit
end = 65 + N
The row must always end at the N-th alphabet.

For N=5, that means end=70 ('F') → so range goes until just before 'F', i.e. up to 'E'.

5. Build the row
python
Copy
Edit
row = "".join(chr(ch) for ch in range(start, end))
range(start, end) → generates all ASCII values from start to end-1.

chr(ch) converts each ASCII number to a character.

"".join(...) concatenates them into a single string with no spaces.

Example (N=5, i=3 → Row 3):

start = 67 ('C'), end = 70 ('F')

Characters: 'C' 'D' 'E'

"".join(...) → "CDE"

6. Print the row
python
Copy
Edit
print(row)
After building one row string, print it.

Then go back to the outer loop for the next row.



Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
11
111
1111

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
5

Sample Output :
1
11
111
1111
11111


In [12]:
n= int(input())
for i in range(1, n+1):
 if i ==1:
  print(1)
 elif i ==2:
  print(11)
 else:
  print('1'*i)

6
1
11
111
1111
11111
111111


# or

In [13]:
N = int(input())

for i in range(1, N+1):
    print("1" * i)


8
1
11
111
1111
11111
111111
1111111
11111111


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
11
202
3003

Detailed explanation ( Input/output format, Notes, Images )
Constraints:
1 <= n <= 50

Output format :
Pattern in N lines

Sample Input :
5

Sample Output :
1
11
202
3003
40004


In [14]:
n= int(input())
for i in range(1, n+1):
  if i ==1:
    print(1)
  if i ==2:
    print(11)
  else:
    print(str(i-1)+'0'*(i-2)+str(i-1))


5
1
00
11
202
3003
40004


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1234
123
12
1

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
5

Sample Output :
12345
1234
123
12
1


In [ ]:
n= int(input())
for i in range(1, n+1):
  for j in range(i, 0, +1)
   print('i')

why my code didnt work because- for the no of rows which is denoted by i it contins 5 elements in the first row
- and j tells us what kind of elements to add

In [15]:
N = int(input())

for i in range(N, 0, -1):   # from N down to 1
    for j in range(1, i+1): # print numbers 1 to i
        print(j, end="")
    print()  # move to next line


4
1234
123
12
1


Print the following pattern for the given N number of rows.
Pattern for N = 3
A
BB
CCC

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 26
Sample Input 1:
7
Sample Output 1:
A
BB
CCC
DDDD
EEEEE
FFFFFF
GGGGGGG
Sample Input 2:
6
Sample Output 2:
A
BB
CCC
DDDD
EEEEE
FFFFFF



In [16]:
n= int(input())
for i in range(1, n+1):
  ch= chr(65+i-1)
  print(ch*i)


5
A
BB
CCC
DDDD
EEEEE


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
4444
333
22
1

Detailed explanation ( Input/output format, Notes, Images )
Constraints :
0 <= N <= 50

Sample Input 1:
5

Sample Output 1:
55555
4444
333
22
1

Sample Input 2:
6

Sample Output 2:
666666
55555
4444
333
22
1




In [17]:
n= int(input())
for i in range(n, 0, -1):
  print(str(i)*i)

4
4444
333
22
1


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4

The dots represent spaces.
Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50

Sample Input 1:
3

Sample Output 1:
     1
    12
  123

Sample Input 2:
4

Sample Output 2:
     1
    12
  123
1234



In [20]:
n= int(input())
for i in range(1, n+1):
 spaces = ""*(n-i)
 char= "".join(str(j) for j in range(1, i+1))
 print(spaces+char)

6
1
12
123
1234
12345
123456


Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 4
1111
000
11
0

Input format : N (Total no. of rows)
Output format : Pattern in N lines
Sample Input :
5

Sample Output :
11111
0000
111
00
1



In [23]:
n= int(input())
for i in range(n):
 length= n-i
 if i%2==0:
  print('1'* length)
 else:
  print('0'*length)

6
111111
00000
1111
000
11
0


Problem statement
Print the following pattern for a given n.
For eg. N = 6
123456
  23456
    3456
      456
        56
          6
        56
      456
    3456
  23456
123456

Sample Input 1 :
4

Sample Output 1 :
1234
  234
    34
      4
    34
  234
1234


In [2]:
n= int(input())
for i in range(1, n+1):
 print(""*2*(i-1)+ "".join(str(j)for j in range(i, n+1)))
for i in range(n-1, 0, -1):
 print(""*2*(i-1)+ "".join(str(j)for j in range(i, n+1)))

5
12345
2345
345
45
5
45
345
2345
12345


In [6]:
n= int(input())
for i in range(1, n+1):
 print("".join(str(j)for j in range(i, n+1)))
for i in range(n-1, 0, -1):
 print("".join(str(j)for j in range(i, n+1)))

5
12345
2345
345
45
5
45
345
2345
12345


In [3]:
def print_pattern(n):
    # First half
    for i in range(1, n+1):
        print(" " * (2*(i-1)) + "".join(str(x) for x in range(i, n+1)))
    # Second half
    for i in range(n-1, 0, -1):
        print(" " * (2*(i-1)) + "".join(str(x) for x in range(i, n+1)))


In [4]:
n=int(input())

5


In [5]:
print(print_pattern(n))

12345
  2345
    345
      45
        5
      45
    345
  2345
12345
None


See the difference between above two

Just because of a bracket it turned out to be different

Ahh got it 👍 thanks for pointing that out — you’re absolutely right ✅
The **first code does run**, but the pattern is different because of **operator precedence**.

Let’s carefully check:

---

### 🔹 First code

```python
print(""*2*(i-1)+ "".join(str(j) for j in range(i, n+1)))
```

Python evaluates from **left to right**:

* `""*2*(i-1)` → always `""` (empty string).
* So effectively the line becomes:

  ```python
  print("".join(str(j) for j in range(i, n+1)))
  ```

This means → **no spaces at all**, just numbers.

👉 Output for `n=4`:

```
1234
234
34
4
34
234
1234
```

---

### 🔹 Second code

```python
print(" " * (2*(i-1)) + "".join(str(x) for x in range(i, n+1)))
```

Here, `" "` (a space) is multiplied, so indentation grows line by line.

👉 Output for `n=4`:

```
1234
  234
    34
      4
    34
  234
1234
```

---

### 🔹 Why the difference?

* In the first code: `""*2*(i-1)` is **always empty**, so no indentation.
* In the second code: `" "` contains a space, so multiplying creates indentation.

That’s why the first one prints a **straight diamond-like shape without spaces**,
while the second one prints the **indented diamond (correct pattern)**.

---

👉 Would you like me to **draw both patterns for n=6 side by side** so you can visually compare the difference better?


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4

The dots represent spaces.
Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50

Sample Input 1:
3

Sample Output 1:
     1
    12
  123

Sample Input 2:
4

Sample Output 2:
     1
    12
  123
1234


In [8]:
n= int(input())
for i in range(1, n+1):
 space= " "*(n-i)
 num= "".join(str(j)for j in range(i, n+1))
 print(space+num)

4
   1234
  234
 34
4


In [9]:
def print_pattern(N):
    for i in range(1, N+1):
        spaces = " " * (N - i)
        numbers = "".join(str(x) for x in range(1, i+1))
        print(spaces + numbers)

# Example usage
print_pattern(4)


   1
  12
 123
1234


Difference between my understanding of the logic

Aha, I see the confusion 👌 Let’s compare carefully.

---

### 🔹 Your Code

```python
n = int(input())
for i in range(1, n+1):
    space = " " * (n - i)
    num = "".join(str(j) for j in range(i, n+1))
    print(space + num)
```

### 🔹 My Code

```python
for i in range(1, n+1):
    space = " " * (n - i)
    num = "".join(str(j) for j in range(1, i+1))
    print(space + num)
```

---

### 🔑 Difference is in this line:

* **Your code**:

  ```python
  num = "".join(str(j) for j in range(i, n+1))
  ```

  → Prints numbers **from i to n**.

* **Correct code**:

  ```python
  num = "".join(str(j) for j in range(1, i+1))
  ```

  → Prints numbers **from 1 to i**.

---

### 🔹 Example for n=4

#### Your output:

```
   1234   (i=1 → numbers from 1 to 4)
  234     (i=2 → numbers from 2 to 4)
 34       (i=3 → numbers from 3 to 4)
4         (i=4 → numbers from 4 to 4)
```

#### Expected output:

```
   1      (i=1 → numbers from 1 to 1)
  12      (i=2 → numbers from 1 to 2)
 123      (i=3 → numbers from 1 to 3)
1234      (i=4 → numbers from 1 to 4)
```

---

✅ So the **misunderstanding**:
You thought the numbers should go from `i` to `n`, but actually they should go from `1` to `i`.

That’s why your triangle looked **upside-down**, while the expected one grows from left to right.

---

Do you want me to **show both patterns side by side for n=5** so you can clearly see the difference?


Problem statement
Print the following pattern
Pattern for N = 4


Hint
As taught in the video, you just have to modify the code so that instead of printing numbers, it should output stars ('*').

The dots represent spaces.



Detailed explanation ( Input/output format, Notes, Images )
Constraints :
0 <= N <= 50

Sample Input 1 :
3

Sample Output 1 :
  *
  ***
 *****

Sample Input 2 :
4

Sample Output 2 :
   *
   ***
  *****
 *******


In [10]:
n= int(input())
for i in range(1, n+1):
  space= " "*(n-i)
  stars= "*"*(2*i-1)
  print(space+stars)

5
    *
   ***
  *****
 *******
*********


Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 4


The dots represent spaces.



Detailed explanation ( Input/output format, Notes, Images )
Constraints :
0 <= N <= 50

Sample Input 1:
5

Sample Output 1:
          1
          232
         34543
        4567654
       567898765

Sample Input 2:
4

Sample Output 2:
          1
          232
         34543
        4567654


In [11]:
n= int(input())
for i in range(1, n+1):
  space= " "*(n-i)
  asc_nu= "".join(str(j)for j in range(i, 2*i))
  desc_nu= "".join(str(j)for j in range(2*i-2, i-1, -1))
  print(space+asc_nu+desc_nu)

5
    1
   232
  34543
 4567654
567898765


Print the following pattern for the given number of rows.
Note: N is always odd.
Pattern for N = 5


The dots represent spaces.



Detailed explanation ( Input/output format, Notes, Images )
Constraints :
1 <= N <= 49

Sample Input 1:
5

Sample Output 1:
 *
 ***
*****
 ***
  *

Sample Input 2:
3

Sample Output 2:
 *
 ***
  *


In [ ]:
n= int(input())
for i in range(1, n+1):
  upper_half= ' '*(2*i-1)+ "*"*(i, n+1)

In [13]:
n= int(input())
def diamond_of_stars(n):
 mid= (n//2)+1
 for i in range(1,n+1):
   space= mid -i
   stars= "*"*(2*i-1)
   print(space+stars)
 for i in range(n-1, 0, -1):
   space= mid -i
   stars= "*"*(2*i-1)
   print(space+stars)
diamond_of_stars(n)

3


TypeError: unsupported operand type(s) for +: 'int' and 'str'

Aha 🔥 great catch — let’s unpack this.

---

### 🔹 Your Code

```python
space = mid - i
stars = "*" * (2*i - 1)
print(space + stars)
```

---

### 🔑 Problem here

* `space` is currently an **integer** (like 2, 3, …).
* `stars` is a **string** (like `"***"`).
* In Python, you **cannot add int + str** → it throws a `TypeError`.

That’s why it doesn’t work unless you “convert” the spaces into a string.

---

### 🔹 Correct Way

Instead of keeping `space` as an integer, you should **build a string of spaces**:

```python
space = " " * (mid - i)   # string of spaces
stars = "*" * (2*i - 1)   # string of stars
print(space + stars)      # string + string works
```

---

### 🔹 Why do we need this?

* `" " * k` → repeats a space `k` times, giving `"   "` (k spaces).
* If you just use `k` (an integer), Python doesn’t know how to “add” that to a string.

Example:

```python
print(3 + "***")   # ❌ Error (int + str not allowed)
print(" " * 3 + "***")  # ✅ "   ***"
```



In [14]:
n= int(input())
def diamond_of_stars(n):
 mid= (n//2)+1
 for i in range(1,n+1):
   space= " "*(mid -i)
   stars= "*"*(2*i-1)
   print(space+stars)
 for i in range(n-1, 0, -1):
   space= " "*(mid -i)
   stars= "*"*(2*i-1)
   print(space+stars)
diamond_of_stars(n)

4
  *
 ***
*****
*******
*****
 ***
  *


number pattern

Problem statement
Print the following pattern for n number of rows.
Note: each line consist of equal number of characters + spaces. Suppose you are printing xth line for N=n. You need to print 1..x followed by (n-x) spaces, again (n-x) spaces followed by x..1
For eg. N = 5
Sample Input 1 :
4

Sample Output 1 :




In [16]:
n= int(input())
def num_patt(n):
 for i in range(1, n+1):
  p1= "".join(str(j)for j in range(1, i+1))
  space= " "*(n-i)
  p2="".join(str(j)for j in range(i, 0, -1))
  print(p1+space+p2)
num_patt(n)

4
1   1
12  21
123 321
12344321


Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 4
  1
  212
 32123
4321234

Input format : N (Total no. of rows)
Output format : Pattern in N lines
Sample Input :
5

Sample Output :
   1
   212
  32123
 4321234
543212345


In [18]:
n=int(input())
for i in range(1, n+1):
 space= " "*(n-i)
 desc= "".join(str(j)for j in range(i, 0, -1))
 asc= "".join(str(j)for j in range(2, i+1))
 print(space+desc+asc)

4
   1
  212
 32123
4321234


Problem statement
Print the following pattern for the given number of rows.
Assume N is always odd.
Note : There is space after every star. Pattern for N = 7
*
 * *
   * * *
     * * * *
   * * *
 * *
*

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
11

Sample Output :
*
 * *
   * * *
     * * * *
       * * * * *
         * * * * * *
       * * * * *
     * * * *
   * * *
 * *
*



In [20]:
n=int(input())
m= n//2+1
for i in range(1, m+1):
 space=" "*(i-1)
 st1= "*"*(i)
 print(space+st1)
for i in range(m-1, 0, -1):
 space=" "*(i-1)
 st2= "*"*(i)
print(space+st2)

4
*
 **
  ***
*


In [24]:
N=int(input())
def diamond_pattern(N):
    mid = N // 2 + 1

    # Top half including middle
    for i in range(1, mid + 1):
        spaces = " " * (i - 1)
        stars = "* " * i
        print(spaces + stars)

    # Bottom half
    for i in range(mid - 1, 0, -1):
        spaces = " " * (i - 1)
        stars = "* " * i
        print(spaces + stars)

diamond_pattern(N)


4
* 
 * * 
  * * * 
 * * 
* 


Problem statement
Print the following pattern
Pattern for N = 4
*000*000*
0*00*00*0
00*0*0*00
000***000

Detailed explanation ( Input/output format, Notes, Images )
Sample Input 1 :
3

Sample Output 1 :
*00*00*
0*0*0*0
00***00

Sample Input 2 :
5

Sample Output 2 :
*0000*0000*
0*000*000*0
00*00*00*00
000*0*0*000
0000***0000


In [27]:
n=int(input())
width= 2*n+1
for i in range(1, n+1):
 row=[] # Initialize row as an empty list
 for j in range(width):
  if j==i or j==n or j==(2*n-i):
    row.append('*')
  else:
    row.append('0')
 print("".join(row)) # Print the row after building it

3
0*0*0*0
00***00
000*000


In [28]:
n = int(input().strip())
width = 2 * n + 1

for i in range(n):            # i = 0 .. n-1
    row = []
    for j in range(width):    # j = 0 .. 2n
        if j == i or j == n or j == 2*n - i:
            row.append('*')
        else:
            row.append('0')
    print(''.join(row))


3
*00*00*
0*0*0*0
00***00


differece between the above two code- **difference between range(1, n+1) and range(n)**




Ahh good that you asked this specifically for **the above star/zero pattern question** 👍

Let’s carefully connect it:

---

### 🔹 Problem recap (for `n = 3`):

```
*00*00*
0*0*0*0
00***00
```

---

### Code structure you saw:

```python
for i in range(1, n+1):   # (1 … n)
    for j in range(1, n+1):
        ...
```

Here, `i` is the **row number** (1-based), and `j` is the **column number** (also 1-based).
We decide where to put `"*"` depending on positions like `(i, j)`.

---

### If we use `range(1, n+1)`:

* Rows go from **1 to n** (matches the way problem statement is written: row 1, row 2, etc).
* Columns go from **1 to n+1** or similar (depending on logic).
* Conditions like

  ```python
  if j == i or j == n-i+1 or ...
  ```

  work naturally, because they are **1-based math conditions**.

---

### If we use `range(n)`:

* Rows go from **0 to n-1**.
* Columns go from **0 to n-1**.
* Then, the same conditions **won’t match** the intended positions, because the formula is off by one.
  Example:

  * In row 1 (1-based), the first `*` should appear when `j == 1`.
  * But if you’re looping with 0-based, that condition becomes `j == 0`.

So you would need to **shift your conditions**:

```python
if j == i or j == n-i+1   # (for 1-based)
```

becomes

```python
if j == i or j == n-1-i   # (for 0-based)
```

---

### ✅ Conclusion:

Both `range(1, n+1)` and `range(n)` can work.

* If you pick `range(1, n+1)` → the math is **natural, matches the question**.
* If you pick `range(n)` → you must **adjust all conditions** to be 0-based.

---

Do you want me to **show both versions of the code (1-based and 0-based)** for your star/zero pattern so you can see the difference side by side?


We used:

python
Copy
Edit
n = int(input().strip())
The reason is .strip() removes leading and trailing whitespace or newline characters from the input string.

Why it’s used here?
When you enter input in coding platforms (like HackerRank, Coding Ninjas, LeetCode, etc.), sometimes the input may contain:

Extra spaces

A trailing newline ("\n")

If you directly write int(input()), it usually still works, because int() ignores spaces around numbers (" 5 " → 5).

But strip() makes it safer and prevents issues if input has extra unwanted spaces.



# Ques
Problem statement
Print the following pattern for a given n.
For eg. N = 6
123456
  23456
    3456
      456
        56
          6
        56
      456
    3456
  23456
123456

Sample Input 1 :
4

Sample Output 1 :
1234
  234
    34
      4
    34
  234
1234



In [2]:
n= int(input())
for i in range(1, n+1):
 spaces= " "*(i-1)
 up= "".join(str(j) for j in range(i, n+1))
 print(spaces+up)
for i in range(n-1, 0, -1):
  spaces= " "*(i-1)
  dp= "".join(str(j) for j in range(i, n+1))
  print(spaces+dp)
print()

4
1234
 234
  34
   4
  34
 234
1234



# Diamond shaped numbers

In [3]:
n = int(input())

# First half
for i in range(1, n+1):
    space = " " * (i-1) * 2   # each space actually counts as 2 so numbers stay aligned
    num = "".join(str(j) for j in range(i, n+1))
    print(space + num)

# Second half
for i in range(n-1, 0, -1):
    space = " " * (i-1) * 2
    num = "".join(str(j) for j in range(i, n+1))
    print(space + num)


4
1234
  234
    34
      4
    34
  234
1234


In [4]:
n = int(input())

# First half
for i in range(1, n+1):
    space = " " * i   # each space actually counts as 2 so numbers stay aligned
    num = "".join(str(j) for j in range(i, n+1))
    print(space + num)

# Second half
for i in range(n-1, 0, -1):
    space = " " * i
    num = "".join(str(j) for j in range(i, n+1))
    print(space + num)


4
 1234
  234
   34
    4
   34
  234
 1234


Problem statement
Print the following pattern for the given number of rows.
Note: N is always odd.
Pattern for N = 5


The dots represent spaces.



Detailed explanation ( Input/output format, Notes, Images )
Constraints :
1 <= N <= 49

Sample Input 1:
5

Sample Output 1:
 *
 ***
*****
 ***
  *

Sample Input 2:
3

Sample Output 2:
 *
 ***
  *


Diamond of stars

For this problem n is always odd

In [8]:
n= int(input())
mid= n//2
for i in range(1, n+1):
  space= " "*(mid-i)
  stars= "*"*(2*i-1)
  print(space+stars)
for i in range(n-1, 0, -1):
  space= " "*(mid-i)
  stars= "*"*(2*i-1)
  print(space+stars)
print()

5
 *
***
*****
*******
*********
*******
*****
***
 *



In [7]:
n = int(input())

def diamond_of_stars(n):
    mid = n // 2   # middle row index (0-based)

    for i in range(n):
        if i <= mid:
            # First half (increasing stars)
            spaces = " " * (mid - i)
            stars = "*" * (2*i + 1)
        else:
            # Second half (decreasing stars)
            spaces = " " * (i - mid)
            stars = "*" * (2*(n-i-1) + 1)
        print(spaces + stars)

diamond_of_stars(n)


5
  *
 ***
*****
 ***
  *


Note the difference between my solution and chatgpt's

Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 4
4444444
4333334
4322234
4321234
4322234
4333334  
4444444

Input format : N (Total no. of rows)
Output format : Pattern in N lines
Sample Input :
3

Sample Output :
33333
32223
32123
32223
33333


In [9]:
n = int(input())

size = 2*n - 1   # total rows & columns in the square
for i in range(size):
    for j in range(size):
        # find minimum distance to any border (top, left, bottom, right)
        min_dist = min(i, j, size-1-i, size-1-j)

        # number decreases as you go inward
        print(n - min_dist, end="")
    print()


3
33333
32223
32123
32223
33333


how did this work? didnt understand how the pattern worked

Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 5
1    2   3    4   5
 11   12  13   14  15
 21   22  23   24  25
 16   17  18   19  20
 6    7    8   9   10

Input format : N (Total no. of rows)
Output format : Pattern in N lines
Sample Input :
4

Sample Output :
1  2  3  4
 9 10 11 12
13 14 15 16
 5  6  7  8



In [10]:
n=int(input())
matrix= [[0]*n for _ in range(n)]
num=1
for j in range(n):
  for i in range(n):
    matrix[i][j]=num
    num+=1

for row in matrix:
 print("".join(str(x) for x in row))

3
147
258
369


In [11]:
n = int(input())

# create an empty matrix
matrix = [[0]*n for _ in range(n)]

num = 1
for j in range(n):           # loop through columns
    for i in range(n):       # loop through rows
        matrix[i][j] = num   # fill column-wise
        num += 1

# print the matrix row by row
for row in matrix:
    print(" ".join(str(x) for x in row))


3
1 4 7
2 5 8
3 6 9


Perfect 🙌 let’s simulate **step by step filling** of the matrix so you can clearly see how numbers go in.

We’ll take the example **N = 4**.

---

### 🔹 Step 1: Empty Matrix

Initially after

```python
matrix = [[0]*4 for _ in range(4)]
```

we have:

```
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
```

---

### 🔹 Step 2: Filling Numbers (column by column)

We use:

```python
for j in range(n):   # column loop
    for i in range(n):   # row loop
        matrix[i][j] = num
        num += 1
```

---

#### ✅ Column `j = 0`

* `i=0` → put `1`
* `i=1` → put `2`
* `i=2` → put `3`
* `i=3` → put `4`

Matrix now:

```
1 0 0 0
2 0 0 0
3 0 0 0
4 0 0 0
```

---

#### ✅ Column `j = 1`

* `i=0` → put `5`
* `i=1` → put `6`
* `i=2` → put `7`
* `i=3` → put `8`

Matrix now:

```
1  5 0 0
2  6 0 0
3  7 0 0
4  8 0 0
```

---

#### ✅ Column `j = 2`

* `i=0` → put `9`
* `i=1` → put `10`
* `i=2` → put `11`
* `i=3` → put `12`

Matrix now:

```
1  5  9  0
2  6 10  0
3  7 11  0
4  8 12  0
```

---

#### ✅ Column `j = 3`

* `i=0` → put `13`
* `i=1` → put `14`
* `i=2` → put `15`
* `i=3` → put `16`

Final matrix:

```
1  5  9 13
2  6 10 14
3  7 11 15
4  8 12 16
```

---

### 🔹 Step 3: Print Row by Row

When we print each row:

```
1  5  9 13
2  6 10 14
3  7 11 15
4  8 12 16
```

---

👉 BUT in the problem statement output for `N=4` is:

```
1  2  3  4
9 10 11 12
13 14 15 16
5  6  7  8
```

This means we don’t just **fill straight column-wise**, but we must **rearrange rows** (like a snake pattern).

So, after filling column-wise, we actually **rotate rows** so they match the required snake-like format.

---

⚡ Would you like me to **rewrite the code to directly produce the problem’s expected output** (without needing to rearrange later)?


Ah! Now I see your exact confusion 😄 — you’re thinking:

> “If `i` and `j` are both in `range(n)` → how can the numbers go beyond `n`? Like 5, 6, 7…?”

Let’s clarify carefully.

---

### 1️⃣ What `range(n)` does

```python
for i in range(n):  # 0,1,2,...,n-1
    for j in range(n):
```

* `i` goes from `0` to `n-1` → **row index**
* `j` goes from `0` to `n-1` → **column index**
* These are just **indexes**, not the numbers we print.

---

### 2️⃣ Where the actual numbers come from

```python
num = 1
matrix[i][j] = num
num += 1
```

* `num` **starts at 1**, and **increases by 1 every time we fill a cell**.
* It is **independent of `i` and `j`**.

Even though `i` and `j` only go from `0` to `n-1`:

* There are `n * n` cells in total.
* Each time we visit a cell, we assign `num` and then increment `num`.
* So numbers naturally go **from 1 up to n\*n**.

✅ That’s why for `n=4` we end up with numbers `1` → `16` even though `i` and `j` are only 0–3.

---

### 3️⃣ Visualization for n=4

| i | j | num assigned |
| - | - | ------------ |
| 0 | 0 | 1            |
| 1 | 0 | 2            |
| 2 | 0 | 3            |
| 3 | 0 | 4            |
| 0 | 1 | 5            |
| 1 | 1 | 6            |
| … | … | …            |
| 3 | 3 | 16           |

* `range(n)` just gives **position**.
* `num` is the **value we print**.

---

### ✅ Key takeaway

* `i` and `j` → indexes (0 to n-1)
* `num` → actual value assigned to the cell
* `num` keeps incrementing → can go beyond `n` easily, up to `n*n`

---

If you want, I can **draw a full 4×4 grid and show how `num` fills each cell step by step**, so it’s crystal clear visually.

Do you want me to do that?


Ah, that’s a common question! Let’s break it down carefully.

---

### The code:

```python
matrix = [[0]*n for _ in range(n)]
```

---

### 1️⃣ What this is doing

* This is a **list comprehension** to create a 2D list (matrix) of size `n × n`.
* `[[0]*n for _ in range(n)]` means:

  1. `[0]*n` → creates a list of `n` zeros. Example: if `n=3` → `[0, 0, 0]`.
  2. `for _ in range(n)` → repeats this `n` times to create `n` rows.

So the final result is a **matrix filled with zeros**:

```python
[
 [0, 0, 0],
 [0, 0, 0],
 [0, 0, 0]
]
```

---

### 2️⃣ Why the underscore `_`?

* `_` is a **conventional variable name in Python** when you **don’t actually need the variable**.
* In this case, we just need to **repeat something `n` times**, we don’t care about the loop index.
* You could also write:

```python
matrix = [[0]*n for i in range(n)]
```

* But `i` isn’t used inside the loop, so Python programmers usually write `_` to indicate **“this variable is ignored”**.

---

### ✅ Summary

* `[0]*n` → creates a single row of zeros
* `for _ in range(n)` → repeat to make multiple rows
* `_` → just a placeholder since we don’t need the loop variable

---

If you want, I can also **draw a step-by-step picture showing how this builds a 2D matrix**, which usually makes it super clear.

Do you want me to do that?


Problem statement
Given three values - Start Fahrenheit Value (S), End Fahrenheit value (E) and Step Size (W), you need to convert all Fahrenheit values from Start to End at the gap of W, into their corresponding Celsius values and print the table.
Note: You don't have to write the main function or take input. It has already been taken care of and you need to just print the integer value . Just write the code that prints Fahrenheit to Celsius table in the function itself.
Formula is C = (F - 32) * 5/9

Detailed explanation ( Input/output format, Notes, Images )
Constraints :
0 <= S <= 1000
0 <= E <= 1000
0 <= W <= 1000

Sample Input 1:
0
100
20

Sample Output 1:
0   -17
20  -6
40  4
60  15
80  26
100 37

Sample Input 2:
120
200
40

Sample Output 2:
120 48
160 71
200 93


In [15]:
for f in range(0, 100, 20):
  c= int((f-32)*5/9)
  print(f"{f} {c}")

0 -17
20 -6
40 4
60 15
80 26


Problem statement
Create a function that determines whether a given number N belongs to the Fibonacci sequence. If N is found in the Fibonacci sequence, the function should return true; otherwise, it should return false.
Detailed explanation ( Input/output format, Notes, Images )
Constraints :
0 <= n <= 10^4

Sample Input 1 :
5

Sample Output 1 :
true

Explanation :
Fibonacci sequence begins 0, 1, 1, 2, 3, 5, ... and so on. Since 5 appears in the sequence.

Sample Input 2 :
14

Sample Output 2 :
false    


Tells us if the n exist in fibonacci series

In [18]:
def is_fibonacci(n): # Added colon
 if n==0 or n==1: # Added colon
  return True

 a,b= 0,1
 while b < n: # Corrected While to while and added colon
   a,b= b, a+b
 return n==b # Corrected indentation

In [19]:
n= int(input())
is_fibonacci(n)

6


False

Problem statement
Check whether a given number ’n’ is a palindrome number.
Note :
Palindrome numbers are the numbers that don't change when reversed.
You don’t need to print anything. Just implement the given function.

Example:
Input: 'n' = 51415
Output: true
Explanation: On reversing, 51415 gives 51415.

Detailed explanation ( Input/output format, Notes, Images )
Sample Input 1 :
1032

Sample Output 1 :
false

Explanation Of Sample Input 1:
1032, on being reversed, gives 2301, which is a totally different number.

Sample Input 2 :
121

Sample Output 2 :
true

Explanation Of Sample Input 2:
121, on being reversed, gives 121, which is the same.

Expected time complexity:
The expected time complexity is O(log(n)).

Constraints :
1 <= n <= 10^9
Time Limit: 1 sec


In [1]:
n= int(input())
def check_pal(n):
 rev= int(str(n)[::-1]) # Convert reversed string directly to int
 if n == rev:
  return True
 else:
  return False

check_pal(n)

757


True

Problem statement
You are given an integer 'n'. Return 'true' if 'n' is an Armstrong number, and 'false' otherwise.

An Armstrong number is a number (with 'k' digits) such that the sum of its digits raised to 'kth' power is equal to the number itself. For example, 371 is an Armstrong number because 3^3 + 7^3 + 1^3 = 371.
Detailed explanation ( Input/output format, Notes, Images )
Sample Input 1 :
1

Sample Output 1 :
true

Explanation of Sample Input 1 :
1 is an Armstrong number as, 1^1 = 1.

Sample Input 2 :
103

Sample Output 2 :
false

Sample Input 3 :
1634

Sample Output 3 :
true

Explanation of Sample Input 3 :
1634 is an Armstrong number, as 1^4 + 6^4 + 3^4 + 4^4 = 1634




In [12]:
n=int(input())
def check_armstrong(n):
 for i in n:
  k= len(str(n))
  v= sum(int(i)**k for i in n)
  if v==n:
    return True
  else:
    return False


345


NameError: name 'v' is not defined

In [10]:
check_armstrong(n)

TypeError: 'int' object is not iterable

In [7]:
def is_armstrong(n):
    # Convert number to string to count digits
    k = len(str(n))

    # Calculate sum of digits raised to k
    total = sum(int(digit) ** k for digit in str(n))

    # Check Armstrong condition
    return total == n


In [8]:
n=int(input())
is_armstrong(n)

567


False

corrected my code

Ah, I see where your code is going wrong 👍. Let’s carefully break it down.

You wrote:

```python
n=int(input())
def check_armstrong(n):
 for i in n:   # ❌ Problem here
  k= len(str(n))
  v= sum(int(i)**k for i in n)   # ❌ Another problem here
  if v==n:
    return True
  else:
    return False
```

---

### 🔴 Problems in your code:

1. **`for i in n:` is invalid** → `n` is an integer, and integers are not iterable. You cannot loop directly over `n`.
   ✅ Instead, you should loop over the **string version** of `n` (like `for i in str(n)`).

2. **You don’t need a loop at all here.**
   You’re already using `sum(int(i)**k for i in str(n))`, which handles iteration. The `for i in n:` loop is redundant and wrong.

3. **Indentation of return statements** → Right now, your code returns inside the loop, so it stops after the first digit. It should only return after checking the whole number.

---

### ✅ Corrected code:

```python
def check_armstrong(n):
    k = len(str(n))  # count digits
    v = sum(int(i)**k for i in str(n))  # powered sum of digits
    return v == n
```

---

### 🔹 Example Run:

```python
print(check_armstrong(371))   # True
print(check_armstrong(103))   # False
print(check_armstrong(1634))  # True
```

---

👉 So the **main mistake** in your code is that you tried to loop over `n` (an integer) instead of looping over `str(n)`, and you were returning too early.

Do you want me to also fix your version **using a `for` loop explicitly** instead of `sum(...)` comprehension, so it looks closer to what you first attempted?


In [ ]:
def check_armstrong(n):
    k = len(str(n))  # count digits
    v = sum(int(i)**k for i in str(n))  # powered sum of digits
    return v == n
